File: problems-by-id/problem_0041-first-missing-positive.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function(nums) {
  let i = 0
  while (i < nums.length) {
      const value = nums[i]
      
      if (value !== i + 1 && 1 <= value && value <= nums.length && nums[i] !== nums[value - 1]) {
          nums[i] = nums[value - 1]
          nums[value - 1] = value
      } else {
          i++
      }
  }
  
  for (let i = 0; i < nums.length; i++) {
      if (nums[i] !== i + 1) {
          return i + 1
      }
  }
  
  return nums.length + 1
};

File: problems-by-id/problem_0041_FirstMissingPositive.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        for (var i = 0; i < nums.length; i++) {
            if (
                nums[i] != i + 1 &&
                1 <= nums[i] &&
                nums[i] <= nums.length &&
                nums[i] != nums[nums[i] - 1]
            ) {
                var temp = nums[nums[i] - 1];
                nums[nums[i] - 1] = nums[i];
                nums[i] = temp;
                i--;
            }
        }

        for (var i = 0; i < nums.length; i++) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }

        return nums.length + 1;
    }
}
